### PR TITLE
nixos/doc: Add chapter “necessary system state”

### PR DESCRIPTION
**Pull request by nicoonoclaste**
## Description of changes

Added a chapter documenting the necessary persistent state, as per @RaitoBezarius' suggestion in #268995.

## Things done

- Manual build on platform(s):
  - [x] x86_64-linux
  - [ ] aarch64-linux
- Get all manpage links to resolve
  - [x] systemd: added all manpages w/ a generating script
  - [ ] `pkgs.shadow` manpages
    Upstream does not have a canonical url for manpages (rendered to HTML)
  - [x] Added a script checking all URLs in `doc/manpage-urls.json` are valid
    Removed one broken link that was not used (based on text search through all `.md` files)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by NickCao** [comment]
And I believe `/var/lib/nixos` should also be persistent, to ensure a stable uid/gid for users/groups across reboots.

**Comment by RaitoBezarius** [comment]
cc @lilyinstarlight as I know you have a fairly comprehensive impermanence setup (feel free to nerdsnipe someone else!)

**Comment by roberth** [comment]
You could document this by making each module declare its state locations, and then gather that up. Added benefit is that you can use it with an actual configuration, or even derive more configuration from it.

If a module implementation were to look like the following, it would imply being able to determine both all locations statically (by ignoring `state.<name>.enable`), for docs, as well as determining the locations in use by your actual config.

```
  config = mkMerge [
    (mkIf cfg.enable ...)
    {
      state."my-module" = {
        locations = [
          "/var/lib/my-module";
        ];
        enable = cfg.enable;
      };
      meta.maintainers = ...;
    }
  ];
```

File: nixos/doc/manual/administration/systemd-state.section.md [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+
+## `/var/log/journal/{machine-id}` {#sec-var-journal}
+
+Lastly, {manpage}`systemd-journald(8)` writes the system's journal in binary
+form to `/var/log/journal/{machine-id}`; if (locally) persisting the entire log
+is desired, it is recommended to make all of `/var/log/journal` persistent.
+
+If not, one can set `Storage=volatile` in {manpage}`journald.conf(5)`
+(`services.journald.storage = "volatile";`), which disables journal persistence
```
Comment (jtojnar) [inline nixos/doc/manual/administration/systemd-state.section.md:49]: Link will ensure the option exists:

```suggestion
([`services.journald.storage = "volatile";`](#opt-services.journald.storage)), which disables journal persistence
```

File: nixos/doc/manual/administration/nixos-state.section.md [hunk 2]
```diff
@@ -0,0 +1,28 @@
+# NixOS {#sec-nixos-state}
+
+## `/nix` {#sec-state-nix}
+
+NixOS needs the entirety of `/nix` to be persistent, as it includes:
+- `/nix/store`, which contains all the system's executables, libraries, and supporting data;
+- `/nix/var/nix`, which contains:
+  - the nix daemon's database;
+  - roots whose transitive closure is preserved when garbage-collecting the nix store;
```
Comment (fricklerhandwerk) [inline nixos/doc/manual/administration/nixos-state.section.md:9]:
```suggestion
  - the Nix daemon's database;
  - roots whose transitive closure is preserved when garbage-collecting the Nix store;
```

**Review by fricklerhandwerk — APPROVED** [review]
This is really good, ready to merge from my side.

**Pull request merged**